e2e/fluentd_test.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"bufio"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containerd/containerd/cio"
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+const (
+	fluentdLogDirName = "./../fluentd-logs"
+)
+
+var testFluentd = func() {
+	// These tests are run in serial because we only define one log driver instance.
+	ginkgo.Describe("fluentd shim logger", ginkgo.Serial, func() {
+		ginkgo.It("should send logs to fluentd log driver", func() {
+			args := map[string]string{
+				logDriverTypeKey: fluentdDriverName,
+				containerIdKey:   testContainerId,
+				containerNameKey: testContainerName,
+				"--verbose":      "true",
+			}
+			creator := cio.BinaryIO(*Binary, args)
+			sendTestLogByContainerd(creator, testLog)
+			validateTestLogsInFluentd(fluentdLogDirName, testLog)
+		})
+	})
+}
+
+func validateTestLogsInFluentd(dirName string, testLog string) {
+	// For single test, there are 3 files in Fluentd log dir: "data.<hash>.log", "data.<hash>.log.meta" and "data.log".
+	// For example: "data.b60581c99383f387cfaba1fc90272852e.log", "data.b60581c99383f387cfaba1fc90272852e.log.meta" and "data.log"
+	// "data.<hash>.log" has the logs that the tests sent.
+	// "data.<hash>.log" can have multiple lines of records following time sequence. Here is a sample content with 3 lines.
+	// 2023-09-16T22:54:11+00:00       123456789012    {"source":"stdout","log":"test-e2e-log","container_id":"123456789012","container_name":"test-container"}
+	// 2023-09-16T22:54:30+00:00       123456789012    {"container_id":"123456789012","container_name":"test-container","source":"stdout","log":"test-e2e-log"}
+	// 2023-09-16T22:56:17+00:00       123456789012    {"container_id":"123456789012","container_name":"test-container","source":"stdout","log":"test-e2e-log"}
+	// The following steps retrieves the "log" field of the third string parsed by tab of the last line to validate the tests sent.
```

Comment:
I'm not clear as to why we only care about the last line's log when all 3 have the `"log": "test-e2e-log"` data. Could just be because I'm not familiar with this. Could you either add more explanation to the comment or just respond to this comment for my own knowledge?

Reply:
The log events are following time sequence. Each e2e test case only sends one log event. For now there is only one case so it doesn't matter. In the future, if there are multiple test cases, assuming they are running in serial, getting the last line could match the log the current test case just sent.